[SPARK-29007][STREAMING][MLLIB][TESTS] Enforce not leaking SparkContext in tests which creates new StreamingContext with new SparkContext

HeartSaVioR opened the pull request:
### What changes were proposed in this pull request?

This patch enforces tests to prevent leaking newly created SparkContext while is created via initializing StreamingContext. Leaking SparkContext in test would make most of following tests being failed as well, so this patch applies defensive programming, trying its best to ensure SparkContext is cleaned up.

### Why are the changes needed?

We got some case in CI build where SparkContext is being leaked and other tests are affected by leaked SparkContext. Ideally we should isolate the environment among tests if possible.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Modified UTs.

HeartSaVioR committed:
[SPARK-29007][STREAMING][MLLIB] Enforce not leaking SparkContext in tests which creates new StreamingContext with new SparkContext

HeartSaVioR committed:
Remove redundant

SparkQA commented:
**[Test build #110231 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110231/testReport)** for PR 25709 at commit [`8a66b15`](https://github.com/apache/spark/commit/8a66b150ad18a03cca8b7a66791082e3bcf40801).

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #110233 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110233/testReport)** for PR 25709 at commit [`7fe3b09`](https://github.com/apache/spark/commit/7fe3b09bd94a82b92221bff826eaffb5a06e5e23).

SparkQA commented:
**[Test build #110231 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110231/testReport)** for PR 25709 at commit [`8a66b15`](https://github.com/apache/spark/commit/8a66b150ad18a03cca8b7a66791082e3bcf40801).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110231/
Test PASSed.

SparkQA commented:
**[Test build #110233 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110233/testReport)** for PR 25709 at commit [`7fe3b09`](https://github.com/apache/spark/commit/7fe3b09bd94a82b92221bff826eaffb5a06e5e23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110233/
Test PASSed.

gaborgsomogyi commented:
I agree with the direction, tests should be as independent as possible.

Maybe such functionality can be wrapped in a trait like `SharedSparkContext`. The actual implementation requires a lot of static function calls which can be forgotten easily on new tests.

vanzin commented:
Well, it's hard to enforce that tests do the right thing (either call the static methods or extend from a common class that does the cleanup); but I'd prefer to follow existing patterns that creating yet another one.

e.g. have a `LocalStreamingContext` trait like there is `LocalSparkContext`

The `withStreamingContext` approach is also good, but to use that everywhere it would mean a lot of changes (mostly related to indentation, though).

HeartSaVioR commented:
Thanks both of you for the valuable feedbacks! I agree trait would be better to not forget and blindly apply to test suites. I'll try to reflect the change.

Personally I'd in favor of try-with-resource-like patterns (withXXXContext methods are similar) so if I need to make a change I'd hope to apply this, but if the test suite already has a field for (Streaming/Spark)Context and cleanup with this, I won't change the approach.

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #110316 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110316/testReport)** for PR 25709 at commit [`0f3a4f4`](https://github.com/apache/spark/commit/0f3a4f4972b218e7c761de3c6c8db2dd87ad6fe2).

HeartSaVioR committed:
Cleans up

HeartSaVioR commented:
Added `LocalStreamingContext` and applied to test suites which clean up StreamingContext in afterXXX. To be consistent (and by intention to isolate SparkContext between test), I just changed the test suites which initialize SparkContext per suite to initialize SparkContext per each test.

SparkQA commented:
**[Test build #110317 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110317/testReport)** for PR 25709 at commit [`fda1c0e`](https://github.com/apache/spark/commit/fda1c0ecc9013a7b394776f6b6e4c60f21a10c82).

SparkQA commented:
**[Test build #110316 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110316/testReport)** for PR 25709 at commit [`0f3a4f4`](https://github.com/apache/spark/commit/0f3a4f4972b218e7c761de3c6c8db2dd87ad6fe2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class StreamingKMeansSuite extends SparkFunSuite with LocalStreamingContext with TestSuiteBase `
  * `class DStreamClosureSuite extends SparkFunSuite with LocalStreamingContext with BeforeAndAfterAll `
  * `class StreamingListenerSuite extends TestSuiteBase with LocalStreamingContext with Matchers `
  * `class InputInfoTrackerSuite extends SparkFunSuite with LocalStreamingContext with BeforeAndAfter `
  * `trait LocalStreamingContext extends BeforeAndAfterEach with BeforeAndAfterAll `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110316/
Test FAILed.

srowen commented on streaming/src/test/scala/org/apache/spark/streamingtest/LocalStreamingContext.scala:
```diff
@@ -0,0 +1,61 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.streamingtest
```
Why `streamingtest` vs `streaming.testutil`, where a similar class exists? We already have `LocalJavaStreamingContext`; can this be combined?

HeartSaVioR replied:
I'm not sure we can combine `LocalStreamingContext` and `LocalJavaStreamingContext`, as

1. LocalJavaStreamingContext relies on `junit`, whereas scala test suites rely on `scalatest`. I'm not sure both can be co-used.
2. LocalStreamingContext has to be a trait to co-use with SparkFunSuite (or its descendant) which many test suites are relying on.
2. It may bring more changes, as `LocalJavaStreamingContext` just initializes JavaStreamingContext in `setUp` whereas Scala test suites initialize StreamingContext when they need it - flexible with how to initialize StreamingContext. If we combine twos into one, the approach `LocalStreamingContext` is doing is the way to go, and we would need to modify existing Java UTs to reflect the change.

As I commented before that we no longer need StreamingTestUtils, I guess we're addressing first question now. Please let me know there's something we can apply more.

srowen replied:
OK that's fine. Can we keep them in the same package at least?

HeartSaVioR replied:
Yes sure. I'll move the package. Thanks for guiding!

SparkQA commented:
**[Test build #110317 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110317/testReport)** for PR 25709 at commit [`fda1c0e`](https://github.com/apache/spark/commit/fda1c0ecc9013a7b394776f6b6e4c60f21a10c82).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110317/
Test FAILed.

srowen commented on streaming/src/main/scala/org/apache/spark/streaming/testutil/StreamingTestUtils.scala:
```diff
@@ -0,0 +1,54 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.streaming.StreamingContext
+
+object StreamingTestUtils extends Logging {
```
No big deal but is it useful to just have these utility methods in `LocalStreamingContext`, in a companion object?

HeartSaVioR replied:
Yeah that sounds better. Actually StreamingTestUtils was introduced to apply the change across different modules - I thought we don't prefer to depend on test jar, but just realized it's already widely used so the original purpose is invalid right now.

srowen reviewed (commented):
The clean up seems good.

HeartSaVioR committed:
Address CI test failure

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #110323 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110323/testReport)** for PR 25709 at commit [`087e1cc`](https://github.com/apache/spark/commit/087e1cc9b3bcd9437d443f77c54cad6d950bde27).

SparkQA commented:
**[Test build #110323 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110323/testReport)** for PR 25709 at commit [`087e1cc`](https://github.com/apache/spark/commit/087e1cc9b3bcd9437d443f77c54cad6d950bde27).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait LocalStreamingContext extends BeforeAndAfterEach with BeforeAndAfterAll `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110323/
Test PASSed.

gaborgsomogyi commented:
I think the question is the granularity of defensiveness. In the Structured Streaming area there were huge execution times because each and every test has created its own `SparkSession`. That's the main reason why `SharedSparkSession` was introduced.

In some cases here we're switching to per test `StreamingContext` creation. Do we have a rough number what does that mean in terms of execution time?

HeartSaVioR commented:
Unfortunately no, I don't have the number, and actually I expect these numbers would fluctuate unless we have some way to correctly measure it.

Both LocalJavaStreamingContext and LocalSparkContext deal with cleanup per test - I agree it would be generally slower then before if test suites deal with cleanup per suite, but that's also obvious that it would be safer. So that sounds me as preference or policy. Let's hear about further opinions - I'm open to roll back and just apply withStreamingContext or even call static function for these suites.

srowen commented:
I'd just compare the runtimes for some key large test suites where this change causes the context to be re-created per test. You can see them in the PR builder and compare to a recent master build. Yes that's also my concern, whether that side effect causes the test time to increase significantly. I also agree it could be worth it for isolation. If we're adding a minute or so? hm, OK. If it's 10 minutes, we probably need to go back and restore the previous behavior in some cases where it's still correct to share the context across tests.

gaborgsomogyi commented:
As @srowen said there are couple of tests involved and can be checked in PR builder. I don't expect really huge jumps (and don't think rollback is needed) but would be good to speak about measurable numbers.

HeartSaVioR commented:
Yeah I was careful how to run them in my dev env. and make the numbers being consistent, but looking PR builder would be a good idea. Let me see some elapsed time for test suites being touched in this patch.

HeartSaVioR commented on external/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisStreamSuite.scala:
```diff
@@ -71,12 +65,6 @@ abstract class KinesisStreamTests(aggregateTestData: Boolean) extends KinesisFun
 
   override def afterAll(): Unit = {
     try {
```
This test suites changed initializing SparkContext "per suite" to "per test" - checking elapsed times...

HeartSaVioR commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamClosureSuite.scala:
```diff
@@ -29,24 +29,14 @@ import org.apache.spark.util.ReturnStatementInClosureException
  */
-class DStreamClosureSuite extends SparkFunSuite with BeforeAndAfterAll {
-  private var ssc: StreamingContext = null
+class DStreamClosureSuite extends SparkFunSuite with LocalStreamingContext with BeforeAndAfterAll {
```
This test suites changed initializing SparkContext and StreamingContext "per suite" to "per test" - checking elapsed times...

HeartSaVioR commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamScopeSuite.scala:
```diff
@@ -30,28 +30,21 @@ import org.apache.spark.util.ManualClock
-class DStreamScopeSuite extends SparkFunSuite with BeforeAndAfter with BeforeAndAfterAll {
-  private var ssc: StreamingContext = null
-  private val batchDuration: Duration = Seconds(1)
+class DStreamScopeSuite
```
This test suites changed initializing SparkContext and StreamingContext "per suite" to "per test" - checking elapsed times...

HeartSaVioR commented on streaming/src/test/scala/org/apache/spark/streaming/ReceiverInputDStreamSuite.scala:
```diff
@@ -30,15 +30,10 @@ import org.apache.spark.streaming.receiver.{BlockManagerBasedStoreResult, Receiv
-      super.afterAll()
-    }
-  }
+class ReceiverInputDStreamSuite
```
This test suites changed cleaning up StreamingContext "per suite" to "per test" - checking elapsed times...

HeartSaVioR commented:
`#110323` is the build number this patch is activated. Let me find the others and compare with.

HeartSaVioR commented:
Here's the comparison between other builds (emphasized `#110323 ` via rectangle):

![SPARK-29007-elapsed-time-comparison-dstream-closure-suite](https://user-images.githubusercontent.com/1317309/64532285-25ffb100-d34c-11e9-8dc0-56f8be8ec964.png)
![SPARK-29007-elapsed-time-comparison-dstream-scope-suite](https://user-images.githubusercontent.com/1317309/64532286-25ffb100-d34c-11e9-8e9a-096866152397.png)
![SPARK-29007-elapsed-time-comparison-kinesis](https://user-images.githubusercontent.com/1317309/64532287-26984780-d34c-11e9-901e-fae48b986e3c.png)
(Showing N mins are only valid - some tests are disabled if flag is turned off.)

![SPARK-29007-elapsed-time-comparison-receiver-input-dstream-suite](https://user-images.githubusercontent.com/1317309/64532288-26984780-d34c-11e9-8980-3953def194ee.png)

Looks like it might be able to contribute couple of seconds (or even more) but it seems to be small compared to the fluctuation among builds.

gaborgsomogyi commented on streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala:
```diff
@@ -0,0 +1,91 @@
+    }
+  }
+
+  def ensureNoActiveSparkContext(ssc: StreamingContext): Unit = {
```
This function does a little bit more + I have the feeling this can be merged with `stop`.
```
  def stop(ssc: StreamingContext, stopSparkContext: Boolean): Unit = {
    if (ssc != null) {
      ssc.stop(stopSparkContext)
    }
    ensureNoActiveSparkContext()
  }
```
BTW, why are these functions are belonging to the object? There are transient vars for `ssc` and `stopSparkContext`.

HeartSaVioR replied:
I'll inline this to `stop()` as there's no usage other than `stop()`.

Regarding placing them to the object instead of trait, it can be also used from outside of LocalStreamingContext trait. `LocalSparkContext` also exposes similar thing. I'll modify the code shortly - it will help deduplicate the code outside of `LocalSparkContext`.

gaborgsomogyi commented:
I've had a look and the time consumption numbers looks good, no significant increase so this part is fine.

HeartSaVioR committed:
Reflect review comment

SparkQA commented:
**[Test build #110366 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110366/testReport)** for PR 25709 at commit [`1bd2b54`](https://github.com/apache/spark/commit/1bd2b54bee1088da870bbfd5fec7726ba5249a9c).

SparkQA commented:
**[Test build #110366 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110366/testReport)** for PR 25709 at commit [`1bd2b54`](https://github.com/apache/spark/commit/1bd2b54bee1088da870bbfd5fec7726ba5249a9c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110366/
Test PASSed.

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala:
```diff
@@ -0,0 +1,89 @@
+      }
+    } catch {
+      case e: Exception =>
+        logError("Error stopping StreamingContext", e)
```
Shouldn't this cause the test to fail? Or are there errors that can occur here that are outside of our control?

HeartSaVioR replied:
That's basically migrated from existing `withStreamingContext` so honestly I thought it's preference (like error on cleanup shouldn't affect test result, saying opposite).

Given we ensure no active spark context is available even exception is thrown, I think impact of not swallowing exception here is restricted to single test failure, which is good. I'll just let it throw exception.

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/LocalStreamingContext.scala:
```diff
@@ -0,0 +1,89 @@
+          sc.stop()
+        } catch {
+          case e: Throwable =>
+            logWarning("Exception trying to stop SparkContext, clear active SparkContext...", e)
```
Same question here. Seems like something that should make the tests fail.

HeartSaVioR replied:
OK. I'll rethrow exception after clearing active context. Thanks for suggestion.

vanzin reviewed (commented):
There are now 3 different ways to run code before and after tests:

- BeforeAndAfterEach inherited from LocalStreamingContext
- BeforeAndAfter
- TestSuiteBase.before/afterFunction

Could you make everybody use BeforeAndAfterEach?

HeartSaVioR committed:
Reflect review comments

HeartSaVioR commented:
Addressed modified suites in this patch to use BeforeAndAfterEach. I didn't check other suites as the scope of PR is getting bigger.

SparkQA commented:
**[Test build #110388 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110388/testReport)** for PR 25709 at commit [`d2ed050`](https://github.com/apache/spark/commit/d2ed050cd36ea7b096d376bc0bebd5f249047687).

SparkQA commented:
**[Test build #110388 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110388/testReport)** for PR 25709 at commit [`d2ed050`](https://github.com/apache/spark/commit/d2ed050cd36ea7b096d376bc0bebd5f249047687).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `trait LocalStreamingContext extends BeforeAndAfterEach `
  * `trait TestSuiteBase extends SparkFunSuite with BeforeAndAfterEach with Logging `
  * `class InputInfoTrackerSuite extends SparkFunSuite with LocalStreamingContext `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110388/
Test PASSed.

gaborgsomogyi reviewed (approved):
LGTM from my side.

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamScopeSuite.scala:
```diff
@@ -30,28 +30,23 @@ import org.apache.spark.util.ManualClock
+  extends SparkFunSuite
+  with LocalStreamingContext
+  with BeforeAndAfter
+  with BeforeAndAfterAll {
```
BeforeAndAfterAll doesn't seem to be used?

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamScopeSuite.scala:
```diff
@@ -30,28 +30,23 @@ import org.apache.spark.util.ManualClock
   }
 
-  before { assertPropertiesNotSet() }
-  after { assertPropertiesNotSet() }
```
You didn't add this back in a `afterEach` block.

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/ExecutorAllocationManagerSuite.scala:
```diff
@@ -19,26 +19,25 @@ package org.apache.spark.streaming.scheduler
-class ExecutorAllocationManagerSuite extends SparkFunSuite
-  with BeforeAndAfter with BeforeAndAfterAll with MockitoSugar with PrivateMethodTester {
+class ExecutorAllocationManagerSuite extends TestSuiteBase
+  with BeforeAndAfterEach with MockitoSugar with PrivateMethodTester {
```
`TestSuiteBase` extends `BeforeAndAfterEach` so adding it again seems redundant.

vanzin commented on streaming/src/test/scala/org/apache/spark/streaming/DStreamScopeSuite.scala:
```diff
@@ -30,28 +30,23 @@ import org.apache.spark.util.ManualClock
+class DStreamScopeSuite
+  extends SparkFunSuite
+  with LocalStreamingContext
+  with BeforeAndAfter
```
Not needed / wanted?

HeartSaVioR replied:
I guess it's not needed. Will remove.

HeartSaVioR committed:
Reflect review comments

SparkQA commented:
**[Test build #110450 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110450/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).

SparkQA commented:
**[Test build #110450 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110450/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110450/
Test FAILed.

HeartSaVioR commented:
Failed tests:

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110450/testReport/org.apache.spark.sql.hive.execution/ObjectHashAggregateSuite/randomized_aggregation_test____with_partial___unsafe__with_distinct____with_grouping_keys___with_empty_input/

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110450/testReport/org.apache.spark.sql.hive.execution/ObjectHashAggregateSuite/randomized_aggregation_test____with_partial___unsafe__with_distinct____with_grouping_keys___with_non_empty_input/

Both are not relevant.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #110460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110460/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).

SparkQA commented:
**[Test build #110460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110460/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110460/
Test FAILed.

HeartSaVioR commented:
Failed test:

https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110460/testReport/org.apache.spark.streaming.kinesis/WithoutAggregationKinesisStreamSuite/failure_recovery/

and...

```
[info] HiveSQLViewSuite:
Exception in thread "Thread-195" java.io.EOFException
	at java.io.ObjectInputStream$BlockDataInputStream.peekByte(ObjectInputStream.java:2960)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1540)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1$React.react(Framework.scala:818)
	at org.scalatest.tools.Framework$ScalaTestRunner$Skeleton$1.run(Framework.scala:807)
	at java.lang.Thread.run(Thread.java:748)
Attempting to post to Github...
[error] running /home/jenkins/workspace/SparkPullRequestBuilder/build/sbt -Phadoop-2.7 -Pkubernetes -Phadoop-cloud -Phive -Phive-thriftserver -Pyarn -Pspark-ganglia-lgpl -Pkinesis-asl -Pmesos -Dtest.exclude.tags=org.apache.spark.tags.ExtendedHiveTest,org.apache.spark.tags.ExtendedYarnTest test ; process was terminated by signal 9
```
not relevant.

HeartSaVioR commented:
retest this, please

SparkQA commented:
**[Test build #110464 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110464/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).

SparkQA commented:
**[Test build #110464 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/110464/testReport)** for PR 25709 at commit [`d12b28e`](https://github.com/apache/spark/commit/d12b28ee23a8d5e2ebf71b7ef85fbe50798f8c1c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/110464/
Test PASSed.

vanzin commented:
(Well, I forgot to actually commit the import statement cleanup before the merge. Not a biggie though.)

HeartSaVioR commented:
Thanks all for reviewing and merging!

dongjoon-hyun commented on mllib/pom.xml:
```diff
@@ -55,6 +55,13 @@
+      <version>${project.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
```
Ur, this duplicates the existing line 119 and 125 and maven complains due to this. I'll make a follow-up~
```
~/P/S/mllib:master$ mvn clean package -DskipTests
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for org.apache.spark:spark-mllib_2.12:jar:3.0.0-SNAPSHOT
[WARNING] 'dependencies.dependency.(groupId:artifactId:type:classifier)' must be unique: org.apache.spark:spark-streaming_${scala.binary.version}:test-jar -> duplicate declaration of version ${project.version} @ line 119, column 17
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
```

dongjoon-hyun replied:
https://github.com/apache/spark/pull/25783 is on the way~

srowen replied:
Oops, good catch.

HeartSaVioR replied:
Thanks for finding and fixing quickly!
